File: src/test/java/net/scriptgate/masonry/transition/TransitionTrace.java
package masonry.transition;

import net.scriptgate.common.Color3f;
import net.scriptgate.common.Point;
import net.scriptgate.engine.Renderer;
import net.scriptgate.masonry.api.Transition;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class TransitionTrace {

    private Transition transition;
    private List<Point> trace;
    private Color3f color;

    private TransitionTrace(Transition transition, Color3f color) {
        this.transition = transition;
        this.trace = new ArrayList<>();
        this.color = color;
    }

    public boolean onTick(double elapsedTime) {
        if (!isCompleted()) {
            return trace.add(transition.getLocationAt(elapsedTime));
        }
        return false;
    }

    public void render(Renderer renderer, Consumer<Point> pointRenderer) {
        renderer.setColor(color);
        trace.forEach(pointRenderer);
    }

    public static TransitionTrace trace(Transition transition, Color3f color) {
        return new TransitionTrace(transition, color);
    }

    public boolean isCompleted() {
        return transition.isCompleted();
    }
}

File: src/test/kotlin/net/scriptgate/masonry/transition/TransitionTrace.kt
package net.scriptgate.masonry.transition

import net.scriptgate.common.Color3f
import net.scriptgate.common.Point
import net.scriptgate.engine.Renderer
import net.scriptgate.masonry.api.Transition
import java.util.*

class TransitionTrace private constructor(private val transition: Transition, private val color: Color3f) {
    private val trace: MutableList<Point>

    val isCompleted: Boolean
        get() = transition.isCompleted

    init {
        this.trace = ArrayList()
    }

    fun onTick(elapsedTime: Double): Boolean {
        return if (!isCompleted) {
            trace.add(transition.getLocationAt(elapsedTime))
        } else false
    }

    fun render(renderer: Renderer, pointRenderer: (Point) -> Unit) {
        renderer.setColor(color)
        trace.forEach(pointRenderer)
    }

    companion object {

        fun trace(transition: Transition, color: Color3f): TransitionTrace {
            return TransitionTrace(transition, color)
        }
    }
}
